README.md
[//]: # "This README.md file is auto-generated, all changes to this file will be lost."
[//]: # "To regenerate it, use `python -m synthtool`."
<img src="https://avatars2.githubusercontent.com/u/2810941?v=3&s=96" alt="Google Cloud Platform logo" title="Google Cloud Platform" align="right" height="96" width="96"/>

# [Cloud Media Translation: Node.js Client](https://github.com/googleapis/nodejs-media-translation)

[![release level](https://img.shields.io/badge/release%20level-beta-yellow.svg?style=flat)](https://cloud.google.com/terms/launch-stages)
[![npm version](https://img.shields.io/npm/v/@google-cloud/media-translation.svg)](https://www.npmjs.org/package/@google-cloud/media-translation)
[![codecov](https://img.shields.io/codecov/c/github/googleapis/nodejs-media-translation/master.svg?style=flat)](https://codecov.io/gh/googleapis/nodejs-media-translation)




Mediatranslation client for Node.js


* [Cloud Media Translation Node.js Client API Reference][client-docs]
* [Cloud Media Translation Documentation][product-docs]
* [github.com/googleapis/nodejs-media-translation](https://github.com/googleapis/nodejs-media-translation)

Read more about the client libraries for Cloud APIs, including the older
Google APIs Client Libraries, in [Client Libraries Explained][explained].

[explained]: https://cloud.google.com/apis/docs/client-libraries-explained

**Table of contents:**


* [Quickstart](#quickstart)
  * [Before you begin](#before-you-begin)
  * [Installing the client library](#installing-the-client-library)
  * [Using the client library](#using-the-client-library)
* [Samples](#samples)
* [Versioning](#versioning)
* [Contributing](#contributing)
* [License](#license)

## Quickstart

### Before you begin

1.  [Select or create a Cloud Platform project][projects].
1.  [Enable billing for your project][billing].
1.  [Enable the Cloud Media Translation API][enable_api].
1.  [Set up authentication with a service account][auth] so you can access the
    API from your local workstation.

### Installing the client library

```bash
npm install @google-cloud/media-translation
```


### Using the client library

```javascript
const fs = require('fs');

// Imports the CLoud Media Translation client library
const {
  SpeechTranslationServiceClient,
} = require('@google-cloud/media-translation');

// Creates a client
const client = new SpeechTranslationServiceClient();

async function quickstart() {
  /**
   * TODO(developer): Uncomment the following lines before running the sample.
   */
  // const filename = 'Local path to audio file, e.g. /path/to/audio.raw';
  // const encoding = 'Encoding of the audio file, e.g. LINEAR16';
  // const sourceLanguage = 'BCP-47 source language code, e.g. en-US';
  // const targetLangauge = 'BCP-47 target language code, e.g. es-ES';

  const config = {
    audioConfig: {
      audioEncoding: encoding,
      sourceLanguageCode: sourceLanguage,
      targetLanguageCode: targetLanguage,
    },
  };

  // First request needs to have only a streaming config, no data.
  const initialRequest = {
    streamingConfig: config,
    audioContent: null,
  };

  const readStream = fs.createReadStream(filename, {
    highWaterMark: 4096,
    encoding: 'base64',
  });

  const chunks = [];
  readStream
    .on('data', chunk => {
      const request = {
        streamingConfig: config,
        audioContent: chunk.toString(),
      };
      chunks.push(request);
    })
    .on('close', () => {
      // Config-only request should be first in stream of requests
      stream.write(initialRequest);
      for (let i = 0; i < chunks.length; i++) {
        stream.write(chunks[i]);
      }
      stream.end();
    });

  const stream = client.streamingTranslateSpeech().on('data', response => {
    const {result} = response;
    if (result.textTranslationResult.isFinal) {
      console.log(
        `\nFinal translation: ${result.textTranslationResult.translation}`
      );
      console.log(`Final recognition result: ${result.recognitionResult}`);
    } else {
      console.log(
        `\nPartial translation: ${result.textTranslationResult.translation}`
      );
      console.log(`Partial recognition result: ${result.recognitionResult}`);
    }
  });


```



## Samples

Samples are in the [`samples/`](https://github.com/googleapis/nodejs-media-translation/tree/master/samples) directory. The samples' `README.md`
has instructions for running the samples.

| Sample                      | Source Code                       | Try it |
| --------------------------- | --------------------------------- | ------ |
| Quickstart | [source code](https://github.com/googleapis/nodejs-media-translation/blob/master/samples/quickstart.js) | [![Open in Cloud Shell][shell_img]](https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/googleapis/nodejs-media-translation&page=editor&open_in_editor=samples/quickstart.js,samples/README.md) |
| Translate_from_file | [source code](https://github.com/googleapis/nodejs-media-translation/blob/master/samples/translate_from_file.js) | [![Open in Cloud Shell][shell_img]](https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/googleapis/nodejs-media-translation&page=editor&open_in_editor=samples/translate_from_file.js,samples/README.md) |
| Translate_from_mic | [source code](https://github.com/googleapis/nodejs-media-translation/blob/master/samples/translate_from_mic.js) | [![Open in Cloud Shell][shell_img]](https://console.cloud.google.com/cloudshell/open?git_repo=https://github.com/googleapis/nodejs-media-translation&page=editor&open_in_editor=samples/translate_from_mic.js,samples/README.md) |



The [Cloud Media Translation Node.js Client API Reference][client-docs] documentation
also contains samples.

## Versioning

This library follows [Semantic Versioning](http://semver.org/).



This library is considered to be in **beta**. This means it is expected to be
mostly stable while we work toward a general availability release; however,
complete stability is not guaranteed. We will address issues and requests
against beta libraries with a high priority.




More Information: [Google Cloud Platform Launch Stages][launch_stages]

[launch_stages]: https://cloud.google.com/terms/launch-stages

## Contributing

Contributions welcome! See the [Contributing Guide](https://github.com/googleapis/nodejs-media-translation/blob/master/CONTRIBUTING.md).

Please note that this `README.md`, the `samples/README.md`,
and a variety of configuration files in this repository (including `.nycrc` and `tsconfig.json`)
are generated from a central template. To edit one of these files, make an edit
to its template in this
[directory](https://github.com/googleapis/synthtool/tree/master/synthtool/gcp/templates/node_library).

## License

Apache Version 2.0

See [LICENSE](https://github.com/googleapis/nodejs-media-translation/blob/master/LICENSE)

[client-docs]: https://googleapis.dev/nodejs/mediatranslation/latest
[product-docs]: https://cloud.google.com/mediatranslation/docs/
[shell_img]: https://gstatic.com/cloudssh/images/open-btn.png
[projects]: https://console.cloud.google.com/project
[billing]: https://support.google.com/cloud/answer/6293499#enable-billing
[enable_api]: https://console.cloud.google.com/flows/enableapi?apiid=mediatranslation.googleapis.com
[auth]: https://cloud.google.com/docs/authentication/getting-started


CHANGELOG.md
# Changelog

## [3.0.0](https://www.github.com/googleapis/nodejs-media-translation/compare/v2.2.3...v3.0.0) (2021-06-24)


### ⚠ BREAKING CHANGES

* Remove unsupported fields: recognition_result and alternative_source_language_codes. docs: Add more comments for supported audio type (#131)

### Bug Fixes

* **deps:** require google-gax v2.12.0 ([#139](https://www.github.com/googleapis/nodejs-media-translation/issues/139)) ([bda81f9](https://www.github.com/googleapis/nodejs-media-translation/commit/bda81f920df258677ae4b24b647ee4c17a0ccbef))
* GoogleAdsError missing using generator version after 1.3.0 ([#147](https://www.github.com/googleapis/nodejs-media-translation/issues/147)) ([735843a](https://www.github.com/googleapis/nodejs-media-translation/commit/735843a9bbeb84be9f61a718271bcd81860a5eaa))
* Remove unsupported fields: recognition_result and alternative_source_language_codes. docs: Add more comments for supported audio type ([#131](https://www.github.com/googleapis/nodejs-media-translation/issues/131)) ([cfaeb1c](https://www.github.com/googleapis/nodejs-media-translation/commit/cfaeb1ce0699eaa372d1ed726bd445177dafe546))
* use require() to load JSON protos ([#141](https://www.github.com/googleapis/nodejs-media-translation/issues/141)) ([1bc887c](https://www.github.com/googleapis/nodejs-media-translation/commit/1bc887cb22014d066e2c382a7f8cf7f518ec903d))

### [2.2.3](https://www.github.com/googleapis/nodejs-media-translation/compare/v2.2.2...v2.2.3) (2021-02-04)


### Bug Fixes

* Added check to see if stream has been destroyed before writing result… ([#107](https://www.github.com/googleapis/nodejs-media-translation/issues/107)) ([5922538](https://www.github.com/googleapis/nodejs-media-translation/commit/592253836803d0ee6002a6e4c4d8a48d219f5d92))

### [2.2.2](https://www.github.com/googleapis/nodejs-media-translation/compare/v2.2.1...v2.2.2) (2020-11-25)


### Bug Fixes

* **browser:** check for fetch on window ([c0245c5](https://www.github.com/googleapis/nodejs-media-translation/commit/c0245c54d8c45f1f3171e9066bc7bab3c4162b69))

### [2.2.1](https://www.github.com/googleapis/nodejs-media-translation/compare/v2.2.0...v2.2.1) (2020-11-07)


### Bug Fixes

* do not modify options object, use defaultScopes ([#94](https://www.github.com/googleapis/nodejs-media-translation/issues/94)) ([5c6d930](https://www.github.com/googleapis/nodejs-media-translation/commit/5c6d930f215a55b876fcda36a1c448aca2b92b8f))

## [2.2.0](https://www.github.com/googleapis/nodejs-media-translation/compare/v2.1.0...v2.2.0) (2020-06-28)


### Features

* move ts target to es2018 from es2016 ([#46](https://www.github.com/googleapis/nodejs-media-translation/issues/46)) ([2e5065b](https://www.github.com/googleapis/nodejs-media-translation/commit/2e5065b0ecb6c38148ec31bf8966617bf7d024d9))
* **samples:** demonstrate real-time conversation translation ([#48](https://www.github.com/googleapis/nodejs-media-translation/issues/48)) ([6331d9d](https://www.github.com/googleapis/nodejs-media-translation/commit/6331d9dd361a003c66ab532f904e1ae298ae296f))


### Bug Fixes

* proper fallback option handling ([e0a5f7d](https://www.github.com/googleapis/nodejs-media-translation/commit/e0a5f7de5bd279e302b597ab73406de651d9937b))

## [2.1.0](https://www.github.com/googleapis/nodejs-media-translation/compare/v2.0.0...v2.1.0) (2020-06-04)


### Features

* adds 'single_utterance' to stream from file sample ([#20](https://www.github.com/googleapis/nodejs-media-translation/issues/20)) ([9455e20](https://www.github.com/googleapis/nodejs-media-translation/commit/9455e207de7ff9c9f72936dd3a908b694eee59fe))


### Bug Fixes

* regen protos and tests, formatting ([#39](https://www.github.com/googleapis/nodejs-media-translation/issues/39)) ([684f355](https://www.github.com/googleapis/nodejs-media-translation/commit/684f355cbecec36c18763b59ddec613c0ea097a9))
* remove eslint, update gax, fix generated protos, run the generator ([#24](https://www.github.com/googleapis/nodejs-media-translation/issues/24)) ([3aca62a](https://www.github.com/googleapis/nodejs-media-translation/commit/3aca62aed3a4e49e4892f9ad29c3d0166c57fa56))
* streaming mic sample ([488c235](https://www.github.com/googleapis/nodejs-media-translation/commit/488c23522b31e26b419c9a4cb5829969e9050928))

## [2.0.0](https://www.github.com/googleapis/nodejs-media-translation/compare/v1.0.0...v2.0.0) (2020-04-05)


### ⚠ BREAKING CHANGES

* The library now supports Node.js v10+. The last version to support Node.js v8 is tagged legacy-8 on NPM.

### Features

* adds media translation file, mic samples ([db91593](https://www.github.com/googleapis/nodejs-media-translation/commit/db915930164484cc016dcb5fbe215d9eef00b412))
* drop node8 support ([#17](https://www.github.com/googleapis/nodejs-media-translation/issues/17)) ([b662dc6](https://www.github.com/googleapis/nodejs-media-translation/commit/b662dc61e43c34b1c56f156eb18d0bbf98170e59))

## 1.0.0 (2020-03-24)


### ⚠ BREAKING CHANGES

* initial library generation (#1)

### Features

* initial library generation ([#1](https://www.github.com/googleapis/nodejs-media-translation/issues/1)) ([33bc8b9](https://www.github.com/googleapis/nodejs-media-translation/commit/33bc8b9469fde7f0d55b7445d2f63e772ed76fee))


samples/translate_from_file.js
// Copyright 2020, Google LLC.
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

'use strict';

/**
 * Translate text from an audio file.
 * @param {string} filename local path to
 * @param {string} encoding the encoding of the audio rate, e.g. Linear16
 * @param {string} sourceLanguage language translating from, as BCP-47 code
 * @param {string} targetLanguage languate translating to, as BCP-47 code
 */
function main(filename, encoding, sourceLanguage, targetLanguage) {
  // [START media_translation_translate_from_file]
  const fs = require('fs');

  // Imports the CLoud Media Translation client library
  const {
    SpeechTranslationServiceClient,
  } = require('@google-cloud/media-translation');

  // Creates a client
  const client = new SpeechTranslationServiceClient();

  async function translate_from_file() {
    /**
     * TODO(developer): Uncomment the following lines before running the sample.
     */
    // const filename = 'Local path to audio file, e.g. /path/to/audio.raw';
    // const encoding = 'Encoding of the audio file, e.g. LINEAR16';
    // const sourceLanguage = 'BCP-47 source language code, e.g. en-US';
    // const targetLangauge = 'BCP-47 target language code, e.g. es-ES';

    const config = {
      audioConfig: {
        audioEncoding: encoding,
        sourceLanguageCode: sourceLanguage,
        targetLanguageCode: targetLanguage,
      },
    };

    // First request needs to have only a streaming config, no data.
    const initialRequest = {
      streamingConfig: config,
      audioContent: null,
    };

    const readStream = fs.createReadStream(filename, {
      highWaterMark: 4096,
      encoding: 'base64',
    });

    const chunks = [];
    readStream
      .on('data', chunk => {
        const request = {
          streamingConfig: config,
          audioContent: chunk.toString(),
        };
        chunks.push(request);
      })
      .on('close', () => {
        // Config-only request should be first in stream of requests
        stream.write(initialRequest);
        for (let i = 0; i < chunks.length; i++) {
          stream.write(chunks[i]);
        }
        stream.end();
      });

    const stream = client.streamingTranslateSpeech().on('data', response => {
      const {result} = response;
      if (result.textTranslationResult.isFinal) {
        console.log(
          `\nFinal translation: ${result.textTranslationResult.translation}`
        );
        console.log(`Final recognition result: ${result.recognitionResult}`);
      } else {
        console.log(
          `\nPartial translation: ${result.textTranslationResult.translation}`
        );
        console.log(`Partial recognition result: ${result.recognitionResult}`);
      }
    });

    // [END media_translation_translate_from_file]
  }
  translate_from_file();
}

main(...process.argv.slice(2));
